File: shared/python/src/shared/itemtype.py
from pebbleobject import pebble

class ItemType (pebble.Pebble):# HasInfoForAutoCrud 
	
	def __init__(self, path, config=None): 
		pebble.Pebble.__init__(self, DocModel.getDoc(path)) 
		self.key = path
		self.config = config
			
	def getFields(self):
		return self.getInheritMember("fields")
	
	def getFunctions (self):
		return self.getInheritMember_Docs("functions")
		
	#fields, functions, typeMaps only
	def getInheritMember(self, member): 
		inheritMember = self.getCreateOnNull(member)
		inherits = self.getRef("inherits")
		if inherits is not None: 
			_base = ItemType(inherits)
			baseInheritMembers = _base.getInheritMember(member)
			for baseInheritMember in baseInheritMembers.getRecords("."):
				inheritMember.set(baseInheritMember.getTagName(), baseInheritMember)
					
		return inheritMember
		
	def getInheritMember_Docs(self, member):
		inheritMember = pebble.Pebble()

		queryItem = pebble.Pebble()
		collectionPath = self.key + "." + member
		queryItem.setRef("path", collectionPath)
		appIndexes = DocModel.getDatasource().doQuery(queryItem)
		for child in appIndexes: 
			childPeb = pebble.Pebble(child)
			inheritMember.setFull(childPeb.getTagName(), childPeb)#need to have uniques!!!

			#even though we show all together, we need to show where it actually resides
			inheritMember.setRef(childPeb.getTagName() + ".ps.memberBaseCollection", self.key + "." + member) 
		

		inherits = self.getRef("inherits")
		if inherits is not None: 
		
			_base = ItemType(inherits)
			baseInheritMembers = _base.getInheritMember_Docs(member)
			for baseInheritMember in baseInheritMembers.getRecords("."):
				inheritMember.setFull(baseInheritMember.getTagName(), baseInheritMember)
		
		return inheritMember
		

	def getField(self, fieldName):
		fields = self.getFields()
		return fields.get(fieldName)


	def	getInnerType(self, relPath, instanceRefs):
		if relPath != "": 
			return self._getInnerType(relPath, instanceRefs)
		else:
			return self

	def _getInnerType(self, relPath, instanceRefs):

		field = self.getInnerField(relPath, instanceRefs)
		if field is not None: 
			return field.get("type")
		
		return None
				
	def getInnerField(self, relPath, instanceRefs):

		if relPath.startswith("."):
			relPath = relPath[1:]
		
		elements = relPath.split('.')
		firstMember = elements[0]
		firstMemberField = self.getFields().get(firstMember)
		if firstMemberField is not None: 
			nextType = firstMemberField.get("type")
			if elements.Length > 1: 
				#do again
				nextTypeStr = nextType.getRef(".")
				nextInstanceRefs = None
				nextRelPath = ""

				#if this is reference
				obj = None
				if nextTypeStr == TypeReferences.REFERENCE:

					#configged
					#string configMeta = firstMemberField.getValue("type.configMeta")
					nextTypeStr = instanceRefs.getRef(firstMember)
					obj = ItemType(nextTypeStr)

					configgedType = nextType.getRef("configgedType")
					if configgedType is not None and configgedType != "": 
						if configgedType.startswith("$reference."):
							configgedType = configgedType.replace("$reference.", "")
							configType = obj.getRef(configgedType)
							if configType is not None: 
								obj = ItemType(configType)
						elif configgedType == "$reference":
							pass
					else:
						pass
						
					nextRelPath = relPath[relPath.find(".") + 1:]
					if instanceRefs is None: 
						nextInstanceRefs = None
					else:
						nextInstanceRefs = instanceRefs.get(firstMember)

				elif nextTypeStr == TypeReferences.ARRAY: 

					arrayFormId = firstMemberField.getRef("type.arrayFormId")
					if arrayFormId is not None: 
						nextRelPath = relPath[relPath.find(".") + 1:]
						nextRelPath = nextRelPath[nextRelPath.find(".") + 1:] #advance it twice
						nextTypeStr = arrayFormId
						if instanceRefs is None: 
							nextInstanceRefs = None
						else:
							nextInstanceRefs = instanceRefs.get(firstMember + "." + elements[1])  #advance it twice
						
					else:
						#why doesnt it have an arrayFormId????
						ii = 1
					
					obj = ItemType(nextTypeStr)
				else:
					nextRelPath = relPath[relPath.find(".") + 1:]
					if instanceRefs == None: 
						nextInstanceRefs = None
					else:
						nextInstanceRefs = instanceRefs.get(firstMember)
					
					obj = ItemType(nextTypeStr)

				return obj.getInnerField(nextRelPath, nextInstanceRefs)

			else:
				#return type
				return firstMemberField
			 
		else:
			#if not defined field, should show error ...
			return None

File: shared/java/src/com/pebblefields/shared/ItemType.java
package com.pebblefields.shared;

import java.io.Serializable;
import java.util.List;

import org.pebblefields.pebbleobject.Pebble;

/**
 * @author ken
 */
public class ItemType extends Pebble implements Serializable{
	public String key;
	public Pebble config;
	public ItemType(String path) {
		super(DocModel.getDoc(path));
        this.key = path;
    }
	public ItemType(String path, Pebble config) {
		super(DocModel.getDoc(path));
        this.key = path;
        
        this.config = config;
	}
	public Pebble getFields() {
    	return this.getInheritMember("fields");
    }
    public Pebble getFunctions () {
    	return this.getInheritMember_Docs("functions");
    }
    //fields, functions, typeMaps only
    private Pebble getInheritMember(String member) {
    	Pebble inheritMember = this.getCreateOnNull(member);
    	String inherits = this.getRef("inherits");
    	if (inherits != null) {
    		ItemType base = new ItemType(inherits);
    		Pebble baseInheritMembers = base.getInheritMember(member);
    		for (Pebble baseInheritMember : baseInheritMembers.getRecords(".")){
    			inheritMember.set(baseInheritMember.getTagName(), baseInheritMember);
    		}
    	}
    	return inheritMember;
    }
    public Pebble getInheritMember_Docs(String member) {
    	Pebble inheritMember = new Pebble();
    	
    	Pebble queryItem = new Pebble();
    	String collectionPath = this.key + "." + member;
		queryItem.setRef("path", collectionPath);
		List<String> appIndexes = DocModel.getDatasource().doQuery(queryItem);
		for (String child : appIndexes) {
			Pebble childPeb = new Pebble(child);
			inheritMember.setFull(childPeb.getTagName(), childPeb);//need to have uniques!!!
			
			//even though we show all together, we need to show where it actually resides
			inheritMember.setRef(childPeb.getTagName() + ".ps.memberBaseCollection", this.key + "." + member); 
		}
    	
    	String inherits = this.getRef("inherits");
    	if (inherits != null) {
    		ItemType base = new ItemType(inherits);
    		Pebble baseInheritMembers = base.getInheritMember_Docs(member);
    		for (Pebble baseInheritMember : baseInheritMembers.getRecords(".")){
    			inheritMember.setFull(baseInheritMember.getTagName(), baseInheritMember);
    		}
    	}
    	return inheritMember;
    }

    public Pebble getField(String fieldName){
    	Pebble fields = this.getFields();
        return fields.get(fieldName);
    }
    
    public Pebble getInnerType(String relPath, Pebble instanceRefs) {
        if (!relPath.isEmpty()) {
            return this._getInnerType(relPath, instanceRefs);
        } else {
            return this;
        }
    }
   
    private Pebble _getInnerType(String relPath, Pebble instanceRefs) {
    	
    	Pebble field = this.getInnerField(relPath, instanceRefs);
    	if (field != null) {
    		return field.get("type");
    	} 
    	return null;
    }
    public Pebble getInnerField(String relPath, Pebble instanceRefs) {
    	    	
    	if (relPath.startsWith(".")){
    		relPath = relPath.substring(1);
    	}
        String[] elements = relPath.split("\\.");
        String firstMember = elements[0];
        Pebble firstMemberField = this.getFields().get(firstMember);
        if (firstMemberField != null) { 
	        Pebble nextType = firstMemberField.get("type");
	        if (elements.length > 1) {// || (nextType.getRef(".").equals(TypeReferences.REFERENCE) && nextType.getRef("configgedType") != null)) { //for refs we need to continue
	            //do again
	            String nextTypeStr = nextType.getRef(".");
	            Pebble nextInstanceRefs = null;
	            String nextRelPath = "";
	            
	          //if this is reference
	            ItemType obj = null;
	        	if (nextTypeStr.equals(TypeReferences.REFERENCE)){// && firstMemberField.get("type.configMeta") != null){
	        		//configged
	        		//String configMeta = firstMemberField.getValue("type.configMeta");
	        		nextTypeStr = instanceRefs.getRef(firstMember);
	        		obj = new ItemType(nextTypeStr);
	        		
	        		String configgedType = nextType.getRef("configgedType");
	        		if (configgedType != null && !configgedType.isEmpty()) {
		        		if (configgedType.startsWith("$reference.")){
		        			configgedType = configgedType.replace("$reference.", "");
		        			String configType = obj.getRef(configgedType);
		            		if (configType != null) {
		            			obj = new ItemType(configType);
		            		}
		        		} else if (configgedType.equals("$reference")){
		        			
		        		}
	        		} else {
	        			
	        		}
	        		
	        			        		 
	        		nextRelPath = relPath.substring(relPath.indexOf(".") + 1);
	        		if (instanceRefs == null) {
	                	nextInstanceRefs = null;
	                } else {
	                	nextInstanceRefs = instanceRefs.get(firstMember);
	                }
	        		
	        	} else if (nextTypeStr.equals(TypeReferences.ARRAY)) {
	        		
	        		String arrayFormId = firstMemberField.getRef("type.arrayFormId");
	        		if (arrayFormId != null) {
	        			nextRelPath = relPath.substring(relPath.indexOf(".") + 1);
	        			nextRelPath = nextRelPath.substring(nextRelPath.indexOf(".") + 1); //advance it twice
	        			nextTypeStr = arrayFormId;
	        			if (instanceRefs == null) {
	                    	nextInstanceRefs = null;
	                    } else {
	                    	nextInstanceRefs = instanceRefs.get(firstMember + "." + elements[1]);  //advance it twice
	                    }
	        		} else {
	        			//why doesnt it have an arrayFormId????
	        			int ii = 1;
	        		}
	        		obj = new ItemType(nextTypeStr);
	        	} else {
	        		nextRelPath = relPath.substring(relPath.indexOf(".") + 1);
	        		if (instanceRefs == null) {
	                	nextInstanceRefs = null;
	                } else {
	                	nextInstanceRefs = instanceRefs.get(firstMember);
	                }
	        		obj = new ItemType(nextTypeStr);
	        	}
	        	 //if (firstMemberTypeRef != null) {
	            
	            return obj.getInnerField(nextRelPath, nextInstanceRefs);
	        //} 
	//        else {
	//            //analyze embedded types
	//            relPath = relPath.substring(relPath.indexOf(".") + 1);
	//            return this._getInnerField(firstMemberField.get("type"), relPath);
	//        }
	                
	        } else {
	            //return type
	            return firstMemberField;
	        } 
        } else {
        	//if not defined field, should show error ...
        	return null;
        }
    }
}
